core/build.gradle

```diff
@@ -1,4 +1,59 @@
+String compat(String src) {
+    if (src.contains('.')) {
+        src.substring(src.lastIndexOf('.')+1)
+    } else {
+        src
+    }
+}
+
+sourceSets {
+    java9 {
+        java {
+            srcDirs = ['src/main/java9']
+        }
+    }
+}
+
+compileJava9Java {
+    sourceCompatibility = 9
+    targetCompatibility = 9
+}
+
+tasks.withType(JavaCompile) {
+    def version = compat(sourceCompatibility)
+    options.compilerArgs.addAll(['--release', version])
+}
+
+tasks.withType(Jar) {
+    into('META-INF/versions/9') {
+        from sourceSets.java9.output
+    }
+    manifest.attributes(
+            'Multi-Release': 'true'
+    )
+}
+
+// Proguard can't handle multi-release jars so we recreate it here.
+tasks.trimShadedJar.doLast {
+    // outjars is a file, so only one jar generated for sure
+    def trimmed = tasks.trimShadedJar.outJarFiles[0].toPath()
+    def copied = trimmed.resolveSibling(trimmed.fileName.toString() + '.bak')
+    java.nio.file.Files.move(trimmed, copied, java.nio.file.StandardCopyOption.REPLACE_EXISTING)
+    ant.zip(destfile: trimmed.toString()) {
```

Comment:
Does `zip` produce a valid JAR? IIRC, a valid JAR's first entry must be `MANIFEST.MF` (or at least some implementations assume so).

Reply:
Thanks didn't know about this - and I found ant's jar command which cleaned this up even more thanks to the perfect `update` parameter :D